### PR TITLE
Add axum binding

### PR DESCRIPTION
**Pull request by andrewwebber**
https://github.com/cloudevents/sdk-rust/issues/149

I found on my travels to this pull request a better understanding of this crates architecture. I think further improvements could be made refactoring warp bindings into a generic http/http_body abstraction for which warp or Axum or tower bindings could be made. However I was not smart enough to accomplish this my spare time.

I hope this is some form useful.

```rust
use axum::prelude::*;
use axum::routing::BoxRoute;
use cloudevents::Event;
use http::StatusCode;

fn app() -> BoxRoute<Body> {
    route(
        "/",
        post(|event: Event| async move {
            println!("received cloudevent {}", &event);
            (StatusCode::OK, event)
        }),
    )
    .boxed()
}

#[tokio::main]
async fn main() {
    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(app().into_make_service())
        .await
        .unwrap();
}
```

```rust
#[tokio::test]
    async fn hello_world() {
        let s =
            axum::Server::bind(&"0.0.0.0:3000".parse().unwrap()).serve(app().into_make_service());
        let h = tokio::spawn(s);

        let time = Utc::now();

        let j = json!({"hello": "world"});
        let req_event = EventBuilderV10::new()
            .id("0001")
            .ty("example.test")
            //TODO this is required now because the message deserializer implictly set default values
            // As soon as this defaulting doesn't happen anymore, we can remove it (Issues #40/#41)
            .time(time)
            .source("http://localhost")
            .data("application/json", j.clone())
            .extension("someint", "10")
            .build()
            .unwrap();

        let client = reqwest::Client::new();
        let resp = client
            .post("http://localhost:3000")
            .event(req_event)
            .unwrap()
            .send()
            .await
            .unwrap()
            .into_event()
            .await
            .unwrap();

        println!("client received event {}", resp);
```

### CODE REVIEW

**Comment by jcrossley3** [comment]
Hey, thanks for this! I'm actually in the process of factoring the common http stuff out in #148. Essentially, your `binding/basichttp` dir is analogous to `binding/http` in that one. It's a bit more ambitious in that it's trying to accommodate actix as well as warp/axum/etc by introducing common traits for headers and response builders. Once I have that one working, I might ask you to rebase this on to that.

**Comment by andrewwebber** [comment]
> Hey, thanks for this! I'm actually in the process of factoring the common http stuff out in #148. Essentially, your `binding/basichttp` dir is analogous to `binding/http` in that one. It's a bit more ambitious in that it's trying to accommodate actix as well as warp/axum/etc by introducing common traits for headers and response builders. Once I have that one working, I might ask you to rebase this on to that.

Sounds great. Happy to rebase at your convenience @jcrossley3

**Comment by andrewwebber** [comment]
Updated to Axum 0.2

```rust
use axum::{
    handler::{get, post},
    routing::BoxRoute,
    Router,
};
use cloudevents::Event;
use http::StatusCode;

fn app() -> Router<BoxRoute> {
    Router::new()
        .route("/", get(|| async { "hello from cloudevents server" }))
        .route(
            "/",
            post(|event: Event| async move {
                println!("received cloudevent {}", &event);
                (StatusCode::OK, event)
            }),
        )
        .boxed()
}

#[tokio::main]
async fn main() {
    axum::Server::bind(&"0.0.0.0:8080".parse().unwrap())
        .serve(app().into_make_service())
        .await
        .unwrap();
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use cloudevents::binding::reqwest::{RequestBuilderExt, ResponseExt};
    use cloudevents::{EventBuilder, EventBuilderV10};
    use serde_json::json;

    #[tokio::test]
    async fn hello_world() {
        let s =
            axum::Server::bind(&"0.0.0.0:3000".parse().unwrap()).serve(app().into_make_service());
        let _ = tokio::spawn(s);

        let time = Utc::now();

        let j = json!({"hello": "world"});
        let req_event = EventBuilderV10::new()
            .id("0001")
            .ty("example.test")
            //TODO this is required now because the message deserializer implictly set default values
            // As soon as this defaulting doesn't happen anymore, we can remove it (Issues #40/#41)
            .time(time)
            .source("http://localhost")
            .data("application/json", j.clone())
            .extension("someint", "10")
            .build()
            .unwrap();

        let client = reqwest::Client::new();
        let resp = client
            .post("http://localhost:3000")
            .event(req_event)
            .unwrap()
            .send()
            .await
            .unwrap()
            .into_event()
            .await
            .unwrap();

        println!("client received event {}", resp);
    }
}
```

**Comment by andrewwebber** [comment]
Hi @jcrossley3 , care appreciate the difficulting in supporting all these diverging web frameworks.

I tried to keep my pull request isolated, small and not change to much of the exciting code base. I am happy however to put some more thought into how we might get this integrated and covered by automated tests.

I am new to Rust so I thought PhantomData was correctly used in this case but happy to learn and be corrected.

**Comment by andrewwebber** [comment]
Hi @jcrossley3 

- rebased given the latest http refactoring changes
- added axum unit tests
- added initial documentation

File: src/binding/axum/mod.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+//! ```
+
+pub mod extract;
+pub mod response;
+mod server_response;
+
+#[cfg(test)]
+mod tests {
+
+    use axum_lib as axum;
```
Comment (dejanb) [inline src/binding/axum/mod.rs:85]: I guess we can move this to `example-projects`

File: src/binding/axum/extract.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        let reason = rejection.0;
+        assert_eq!(reason, StatusCode::BAD_REQUEST)
+    }
+
+    #[tokio::test]
+    async fn axum_test_request_with_full_data() {
+        let time = Utc::now();
+        let j = json!({"hello": "world"});
+
+        let expected = EventBuilderV10::new()
```
Comment (dejanb) [inline src/binding/axum/extract.rs:115]: I just notices that we have still work to do on using actual fixtures in unit tests. I'll jump on this next and then we can change these as well.

File: src/binding/axum/mod.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+//!
+//! ```
+
+pub mod extract;
+pub mod response;
+mod server_response;
+
+#[cfg(test)]
+mod tests {
+
```
Comment (dejanb) [inline src/binding/axum/mod.rs:84]: I guess we can move this to `example-project`

**Review by dejanb — COMMENTED** [review]
Looks good to me on the first look. The comments about moving the example to `examples-projects` and improving tests with fixtures can be done in separate PRs if there's no other issues.

File: src/binding/axum/extract.rs (hunk 2 above)
Comment (andrewwebber) [reply src/binding/axum/extract.rs:115]: Thank you for your feedback.
Would love to learn what that suggestion looks like in your pull request so I can support by example.

File: src/binding/axum/mod.rs (hunk 3 above)
Comment (andrewwebber) [reply src/binding/axum/mod.rs:84]: Agreed.

@dejanb is it part of this projects code of conduct for you to resolve the conversation once satisfied or should I?
Would you like to see the addition to the example projects folder in this pull request?

Comment (jcrossley3) [reply src/binding/axum/mod.rs:84]: I agree with moving this to be an example in this PR, if it's not too much trouble. And our CoC is pretty flexible wrt convo resolutions. I'd say feel free to mark it resolved once you've made that change?

Comment (andrewwebber) [reply src/binding/axum/mod.rs:84]: I am happy to create an example in the example-projects folder.
It is still not useful to have module/binding level integration tests?

Comment (jcrossley3) [reply src/binding/axum/mod.rs:84]: They are useful, yes, so they can remain here and a minimal example can be copied over, perhaps with only a single test just for completeness. We only build the examples, not test them: https://github.com/cloudevents/sdk-rust/blob/master/.github/workflows/rust_tests.yml#L100

Comment (andrewwebber) [reply src/binding/axum/mod.rs:84]: @jcrossley3 axum-example now part of the pull request, located in example-projects

Comment (andrewwebber) [reply src/binding/axum/mod.rs:84]: I made some project wide cargo clippy fixes along the way, hope that is ok.

I would say I am now finished with my changes.

**Review by jcrossley3 — COMMENTED** [review]
This is a great contribution, thanks @andrewwebber !!

File: example-projects/axum-example/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+[package]
```
[the remark this answers is not in the document]
Comment (andrewwebber) [reply example-projects/axum-example/Cargo.toml:1]: Stanza added

File: src/binding/axum/server_response.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            .take()
+            .body(Body::from(bytes))
+            .map_err(|e| crate::message::Error::Other {
+                source: Box::new(e),
+            })
+    }
+    fn finish(&mut self) -> Result<Response<Body>> {
+        self.body(Vec::new())
+    }
+}
```
[the remark this answers is not in the document]
Comment (jcrossley3) [reply src/binding/axum/server_response.rs:29]: Please review #154 at your convenience, @andrewwebber -- thanks!

**Pull request merged**